src/ir/effects.h

```diff
@@ -466,43 +484,61 @@ class EffectAnalyzer {
         return;
       }
 
+      const EffectAnalyzer* targetEffects = nullptr;
+      if (parent.funcEffectsMap) {
+        auto iter = parent.funcEffectsMap->find(curr->target);
+        if (iter != parent.funcEffectsMap->end()) {
+          targetEffects = &iter->second;
+        }
+      }
+
       if (curr->isReturn) {
         parent.branchesOut = true;
+        // When EH is enabled, any call can throw.
+        if (parent.features.hasExceptionHandling() &&
+            (!targetEffects || targetEffects->throws())) {
+          parent.hasReturnCallThrow = true;
+        }
       }
 
-      if (parent.funcEffectsMap) {
-        auto iter = parent.funcEffectsMap->find(curr->target);
-        if (iter != parent.funcEffectsMap->end()) {
-          // We have effect information for this call target, and can just use
-          // that. The one change we may want to make is to remove throws_, if
-          // the target function throws and we know that will be caught anyhow,
-          // the same as the code below for the general path.
-          const auto& targetEffects = iter->second;
-          if (targetEffects.throws_ && parent.tryDepth > 0) {
-            auto filteredEffects = targetEffects;
-            filteredEffects.throws_ = false;
-            parent.mergeIn(filteredEffects);
-          } else {
-            // Just merge in all the effects.
-            parent.mergeIn(targetEffects);
-          }
-          return;
+      if (targetEffects) {
+        // We have effect information for this call target, and can just use
+        // that. The one change we may want to make is to remove throws_, if the
+        // target function throws and we know that will be caught anyhow, the
+        // same as the code below for the general path. We can always filter out
+        // throws for return calls because they are already more precisely
+        // captured by `hasReturnCallThrow`.
+        if (targetEffects->throws_ && (parent.tryDepth > 0 || curr->isReturn)) {
+          auto filteredEffects = *targetEffects;
+          filteredEffects.throws_ = false;
+          parent.mergeIn(filteredEffects);
+        } else {
+          // Just merge in all the effects.
+          parent.mergeIn(*targetEffects);
         }
+        return;
       }
 
       parent.calls = true;
-      // When EH is enabled, any call can throw.
-      if (parent.features.hasExceptionHandling() && parent.tryDepth == 0) {
+      // When EH is enabled, any call can throw. Skip this for return calls
+      // because the throw is already more precisely captured by
+      // `hasReturnCallThrow`.
```

Comment:
```suggestion
      // `branchesOut`.
```

Reply:
`branchesOut` isn't related to the throw effect, though. It's the fact that we set `hasReturnCallThrow` that ensures the throw is eventually reported correctly.

Reply:
Correct, but the context here is removing the throws effect on the next line. It is only valid to remove the throws effect because `branchesOut` indicates "there is a branch from here to some other effects". Those other effects are `hasReturnCallThrow`, but those do not come into play normally - not unless you walk the entire function. So it's the combination of the two, but `branchesOut` is the more immediate thing that "replaces" the throws we are removing, it kind of points to where the effects are at.

Reply:
Consider that we still apply `branchesOut` even if the callee does not throw, though.

I'll update the comment to mention both.

Reply:
We do, yes. But in the `return_call` case that has the additional important meaning of "linking" us to the effects we are removing here. So it explains why the removal of `throws` is valid in that case. That is, `branchesOut` becomes even more significant for return calls given this optimization here, and if we some day refactor `branchesOut` we'd need to consider this place.